### PR TITLE
Spark 3.1 indirect write support

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import org.apache.spark.sql.types.StructType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BigQueryWriteHelper {
+
+  private static final Logger logger = LoggerFactory.getLogger(BigQueryWriteHelper.class);
+
+  private final BigQueryClient bigQueryClient;
+  private final SQLContext sqlContext;
```
[the remark this answers is not in the document]
Comment (VinayLondhe14) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:60]: can you fix this?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            .collect(Collectors.toMap(StructField::name, Function.identity()));
+
+    if (!fieldsToUpdate.isEmpty()) {
+      logger.debug("updating schema, found fields to update: {}", fieldsToUpdate.keySet());
+      TableInfo originalTableInfo = bigQueryClient.getTable(config.getTableIdWithoutThePartition());
+      TableDefinition originalTableDefinition = originalTableInfo.getDefinition();
+      Schema originalSchema = originalTableDefinition.getSchema();
+      Schema updatedSchema =
+          Schema.of(
+              originalSchema.getFields().stream()
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:158]: @sonatype-lift ignore

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    return newField.build();
+  }
+
+  void cleanTemporaryGcsPathIfNeeded() {
+    // TODO(davidrab): add flag to disable the deletion?
+    createTemporaryPathDeleter.ifPresent(IntermediateDataCleaner::deletePath);
+  }
+
+  void verifySaveMode() {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java]: Calling this method now

**Review by VinayLondhe14 — APPROVED** [review]

**Pull request merged**